source/common/event/win32/signal_impl.h

```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+#include "envoy/event/signal.h"
+#include "envoy/network/io_handle.h"
+
+#include "common/event/dispatcher_impl.h"
+#include "common/event/event_impl_base.h"
+#include "common/network/io_socket_handle_impl.h"
+#include "common/singleton/threadsafe_singleton.h"
+
+#include "absl/container/flat_hash_map.h"
+
+namespace Envoy {
+namespace Event {
+
+/**
+ * libevent implementation of Event::SignalEvent.
+ */
+class SignalEventImpl : public SignalEvent {
+public:
+  SignalEventImpl(DispatcherImpl& dispatcher, signal_t signal_num, SignalCb cb);
+
+private:
+  SignalCb cb_;
+  Network::IoHandlePtr read_handle_;
+};
+
+// Windows ConsoleControlHandler does not allow for a context. As a result the thread
+// spawned to handle the console events communicates with the main program with this socketpair.
+// Here we have a map from signal types to IoHandle. When we write to this handle we trigger an
+// event that notifies Envoy to act on the signal.
+using eventBridgeHandlersSingleton =
+    ThreadSafeSingleton<absl::flat_hash_map<signal_t, std::shared_ptr<Network::IoHandle>>>;
```

Comment:
Is there a mechanism that clears ThreadSafeSingleton between tests?

Reply:
No it doesn't but I don't think it matters anyway. There are two reasons for that:

1. We only add new events. This means that the `map`/`array` is bounded in size.
2. There is no way to trigger these signals when you are running inside bazel or even gtest.

The only way to actually go through this code is by spawning envoy out of process. This is because the signals that we are catching are part of the console actually not signals to the process. And both bazel and gtest handle these console signals themselves so the code here is never executed.